### PR TITLE
Refuse min proto version > max proto version (cont.)

### LINKED ISSUE #5127: set_min_proto_version() does not validate that min <= max
The setter combination ``SSL_CTX_set_min_proto_version()`` and ``SSL_CTX_set_max_proto_version()`` as well as  ``SSL_set_min_proto_version()`` and ``SSL_set_max_proto_version()`` accept invalid combinations of min and max protocol. For example it is possible to set a minimum version of TLS 1.2 but a maximum version of TLS 1.0. The helper function ``ssl_check_allowed_versions`` should prevent ``min_version > max_version``.

### PR DESCRIPTION
**Pull request by levitte**
(originally by @tiran)

SSL_CTX_set_min_proto_version() and SSL_CTX_set_max_proto_version() as
well as their SSL equivalents now validate that the mininum protocol
version is lower or requal the maximum protocol version.

Add test cases for min/max protocol API

Closes: https://github.com/openssl/openssl/issues/5127
Signed-off-by: Christian Heimes <christian@python.org>

Also see https://github.com/openssl/openssl/pull/5126

### CODE REVIEW

**Comment by kroeckx** [comment]
Please see #5128

**Comment by kroeckx** [comment]
As in, I think this PR is not what we want in the current state.

**Comment by levitte** [comment]
So OK, but you wanted to keep the test program, right?

**Comment by kroeckx** [comment]
Yes, just the test suite changes.

**Comment by levitte** [comment]
Revert added (I'll squash on merge)

File: test/ssl_ctx_test.c [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    int expected_min;
+    int expected_max;
+} version_test;
+
+static const version_test version_testdata[] = {
+    /* min           max             ok    expected min    expected max */
+    {0,              0,              1, 1, 0,              0},
+    {TLS1_VERSION,   TLS1_2_VERSION, 1, 1, TLS1_VERSION,   TLS1_2_VERSION},
+    {TLS1_2_VERSION, TLS1_2_VERSION, 1, 1, TLS1_2_VERSION, TLS1_2_VERSION},
+    {TLS1_2_VERSION, TLS1_1_VERSION, 1, 0, TLS1_2_VERSION, 0},
```
Comment (kroeckx) [inline test/ssl_ctx_test.c]: I'm not sure why this test passes. I expected setting the max version should in fact work, but fail when you try to connect using it.

**Review by paulidale — APPROVED** [review]
Is still wanted?

**Comment by levitte** [comment]
This was originally made for 1.1.1 (master at the time).  @paulidale, do you approve for this getting cherry-picked to that branch?

**Review by paulidale — APPROVED** [review]
It seems like it would still be worthwhile cherry picking.

**Comment by levitte** [comment]
Merged.

master:
132b5facf8d681db5dfa45828d8b02f1bf5df64b Add test cases for min/max protocol API

1.1.1:
c4ab488399186925c8aaa3678eb1b9c79db656e3 Add test cases for min/max protocol API

Unfortunately, I realised too late that `test/build.info` on master isn't quite right, there's an inclusion missing.  I'll follow up with a build fixup.

**Comment by levitte** [comment]
Another thing no-one seems to have realised is that this test program isn't called from any test recipe.  If someone wants to write a quick recipe or add a line in a suitable already existing one, PR welcome!

**Pull request closed without merging**